tests/keithley_2700.py

```diff
@@ -150,13 +153,241 @@ def test_WHEN_end_channel_range_set_THEN_end_channel_matches_the_set_state(self)
+
+    @contextmanager
+    def _insert_reading(self, reading):
+        self._lewis.backdoor_run_function_on_device("insert_mock_data", [reading])
+        time.sleep(0.5)  # for synchronicity help
+        try:
+            yield
+        finally:
+            pass
+
+    @skip_if_recsim("Cannot use lewis backdoor in recsim")
+    def test_GIVEN_empty_buffer_WHEN_readings_added_to_fill_THEN_buffer_clears_and_index_PVs_correct(self):
+        buffer_size = 10
+        self._set_buffer_size(buffer_size)
+        reads = self._generate_readings(25, 1)
+
+        for i in range(len(reads)):
+            # WHEN
+            with self._insert_reading([reads[i]]):
+                # THEN
+                self.ca.assert_that_pv_is("BUFF:NEXT", (i+1) % buffer_size)
+                self.ca.assert_that_pv_is("INDEX:START", (i+1) % buffer_size)
+
+    @skip_if_recsim("Cannot use lewis backdoor in recsim")
+    def test_GIVEN_empty_buffer_WHEN_new_reads_added_until_buffer_full_THEN_buffer_and_index_pvs_behave_correctly(self):
+        buffer_size = 5
+        self._set_buffer_size(buffer_size)
+        reads = self._generate_readings(6, 5)
+
+        self.ca.assert_that_pv_is("BUFF:NEXT", 0)
+        self.ca.assert_that_pv_is("INDEX:START", 0)
+
+        for i in range(buffer_size - 1):  # -1 so that buffer has 1 free space
+            with self._insert_reading([reads[i]]):
+                self.ca.assert_that_pv_is("BUFF:NEXT", i+1)
+                self.ca.assert_that_pv_is("INDEX:START", i+1)
+        # insert final reading
+        with self._insert_reading([reads[5]]):
+            self.ca.assert_that_pv_is("BUFF:NEXT", 0)
+            self.ca.assert_that_pv_is("INDEX:START", 0)
+
+    @skip_if_recsim("Cannot use lewis backdoor in recsim")
+    def test_GIVEN_buffer_full_WHEN_new_reading_added_THEN_new_reading_added_at_index_0_AND_buffer_still_used(self):
+        buffer_size = 10
+        self._set_buffer_size(buffer_size)
+        reads = self._generate_readings(15, 5)  # 10 greater than buffer capacity
+
+        with self._insert_reading(reads[:buffer_size]):
+            # GIVEN
+            self.ca.assert_that_pv_is("BUFF:NEXT", 0)
+        # WHEN
+        with self._insert_reading([reads[11]]):
+            # THEN
+            self.ca.assert_that_pv_is("BUFF:NEXT", 1)
+        # AND
+        self.ca.assert_that_pv_is("BUFF:CONTROLMODE", "ALW")
+
+    @skip_if_recsim("Cannot use lewis backdoor in recsim")
+    def test_GIVEN_buffer_almost_full_WHEN_multiple_remaining_locations_written_to_THEN_correct_readings_returned(self):
+        self._set_buffer_size(10)
+        reads = self._generate_readings(15, 5)
+        expected_reads = reads[7:10]
+
+        with self._insert_reading(reads[:7]):
+            self.ca.assert_that_pv_is("BUFF:NEXT", 7)
+
+        with self._insert_reading(reads[7:10]):  # This fills the buffer
+            self.ca.assert_that_pv_is("BUFF:NEXT", 0)
+
+        retrieved_readings = self.ca.get_pv_value("BUFF:READ")[:9]
```

Comment:
For the `BUFF:READ` calls, I think it would be useful to have a comment to clarify that you'll return 3 values for each inserted reading.